### PR TITLE
Jk/cumulus 3701 fix null rules

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-3701](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3701)

## Changes

- **CUMULUS-3701**
  - Updated `@cumulus/api` to no longer improperly pass PATCH/PUT null values to Eventbridge rules

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/api/endpoints/rules.js [hunk 1]
```diff
@@ -133,7 +139,7 @@ async function post(req, res) {
  *
  * @param {object} params                   - params object
  * @param {object} params.res               - express response object
- * @param {object} params.oldApiRule        - API 'rule' to update
+ * @param {RuleRecord} params.oldApiRule        - API 'rule' to update
```
Comment (etcart) [inline packages/api/endpoints/rules.js:142]:
```suggestion
 * @param {RuleRecord} params.oldApiRule    - API 'rule' to update
```

File: packages/api/tests/lib/rules/test-rulesHelpers.js
Comment (etcart) [inline packages/api/tests/lib/rules/test-rulesHelpers.js:1]: these tests feel unnecessarily complex and hard to parse. could we check explicitly for the absence of attributes instead of this recursive function? if we'd rather recursively check for no-null, can we pull it outside of a callback or at declare types for objects passed around?

**Review by etcart — COMMENTED** [review]
looks good, one very nit, and one place I think it could be cleaner

**Review by npauzenga — CHANGES_REQUESTED** [review]
This is looking good to me @Jkovarik! Quick suggestion on moving a helper. I don't see other potential problems here but it's tricky with the PUT behavior inserting null values. Is there a case to be made to evaluate our other services (CMR, LZARDS, etc.)?

File: packages/api/lib/rulesHelpers.js [hunk 2]
```diff
@@ -635,14 +639,34 @@ async function invokeRerun(rule) {
...
 /**
- * Updates rule trigger for rule
+ * Recursively removes keys with null values from an RuleRecord.
+ *
+ * This function creates a deep copy of the input object, and then recursively removes all keys
+ * that have null values. The input object is not modified.
+ * @param {any} record - The object to remove null key-values from.
+ * @returns The modified object, with all keys that had null values removed.
+ */
+function removeNullKeyValues(record) {
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/rulesHelpers.js:649]: I was a bit surprised this hadn't been done in lodash, etc.     I'm happy to move it if you feel that's not heading into YAGNI territory.

**Comment by Jkovarik** [comment]
@npauzenga   
"I don't see other potential problems here but it's tricky with the PUT behavior inserting null values. Is there a case to be made to evaluate our other services (CMR, LZARDS, etc.)?"

I'm not sure what you mean - nulled values are essentially key deletions when stored via the REST interface.

File: packages/api/tests/lib/rules/test-rulesHelpers.js
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:1]: Could you give a specific example of what you mean - this comment is a file level comment for the test - I'm assuming you're talking about the tests for the null removal function and/or removing values in this way.  

I don't believe simply checking for queueUrl is sufficient to address the concern - null values are allowed for all fields for the incoming object into the API, however they aren't valid for the written record.      The implementation itself is a fairly straightforward recursion implementation (so much so that I'm a little shocked this wasn't in lodash, common, etc), but *does* require testing of cases.   

We can't strongly type the removal (and I'm reasonably sure we *don't want to*) logic, as the thing validating what the user passed in even matches a rule is `validateRecord`.  

We *could* make updates to the existing code to end-run the need for this by transaction-wrapping the AWS update and using the database object, however that has possible consequences that probably should be scoped into another ticket.

Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:1]: Heh, fail on two counts, use of omitBy seems *far* better and is in common.   Lovely.   I'll update.

File: packages/api/lib/rulesHelpers.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/rulesHelpers.js:649]: Ah, I see, we in fact have just such a thing.   Not sure how I missed that on an initial look-through, but this implementation took almost no time.      I'll remove it and use the common module.

[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lib/rulesHelpers.js:649]: It definitely could be the case that only this one implementation in rules exists but it does seem like the function itself isn't specific to rules. That, to me, makes it a better fit in a common utility, even if it's only ever used here.

Does this really matter? No 😄

**Comment by npauzenga** [comment]
> @npauzenga "I don't see other potential problems here but it's tricky with the PUT behavior inserting null values. Is there a case to be made to evaluate our other services (CMR, LZARDS, etc.)?"
> 
> I'm not sure what you mean - nulled values are essentially key deletions when stored via the REST interface.

Sure, so what I'm having trouble expressing is basically "This was an oversight talking to an external service. Is there potentially another we should be looking for?". This kind of thing is really tough to spot and not necessarily tested as-is so just concerned with 'what else we're missing'.

File: packages/api/lib/rulesHelpers.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/rulesHelpers.js:649]: As it turns out, though, your comment made me notice *we do have such an implementation* :) 

Let me know what you think of  7aaa85bd616

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/rulesHelpers.js]: 7aaa85bd616

File: packages/api/tests/lib/rules/test-rulesHelpers.js
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:1]: An excellent point, thank you for the feedback.    Let me know what you think of 7aaa85bd616

**Comment by Jkovarik** [comment]
@npauzenga  

In the PUT/POST/PATCH endpoints, the external servcices we’re dealing with are eventbridge, kinesis, sns, sqs.    For the latter three, it looks like they’re pulling specific keys from the rule in the external interactions (e.g. rule.value) that will fail if nullified or just missing anyway, else they trigger a consumer lambda that reads the rules from the API at invocation, which is reading from the database records. 


Eventbridge is the only trigger that’s duplicating the rule information in AWS meta, so as to avoid needing a message consumer.   We should probably seriously reconsider that pattern (as part of another PR ) instead of attempting to handle it in the API code.

File: packages/api/lib/rulesHelpers.js [hunk 3]
```diff
@@ -590,8 +593,8 @@ async function addRule(item, payload) {
 /**
  * Checks if record is valid
  *
- * @param {RuleRecord} rule - Rule to check validation
- * @returns {void}          - Returns if record is valid, throws error otherwise
+ * @param {any} rule - Object to validate as a Rule Record validation
+ * @returns {RuleRecord}          - Returns if record is valid, throws error otherwise
```
Comment (etcart) [inline packages/api/lib/rulesHelpers.js:597]:
```suggestion
 * @param {any} rule     - Object to validate as a Rule Record validation
 * @returns {RuleRecord} - Returns if record is valid, throws error otherwise
```

**Comment by etcart** [comment]
I feel good about the issues I've raised now

**Review by npauzenga — APPROVED** [review]
Looks great @Jkovarik. Thanks for thinking through the put/patch behavior!
